packages/perspective-viewer-d3fc/src/js/legend/styling/enableDragging.js

```diff
@@ -0,0 +1,65 @@
+/******************************************************************************
+ *
+ * Copyright (c) 2017, the Perspective Authors.
+ *
+ * This file is part of the Perspective library, distributed under the terms of
+ * the Apache License 2.0.  The full license can be found in the LICENSE file.
+ *
+ */
+
+import * as d3 from "d3";
+import {isElementOverflowing} from "../../utils/utils";
+import {getChartElement} from "../../plugin/root";
+
+export function enableDragging(element) {
+    const node = element.node();
+    node.style.cursor = "move";
+
+    const drag = d3.drag().on("drag", function() {
+        overrideDefaultResizingBehaviour(node);
+        const [offsetX, offsetY] = enforceContainerBoundaries(this, d3.event.dx, d3.event.dy);
+        this.style.left = `${this.offsetLeft + offsetX}px`;
+        this.style.top = `${this.offsetTop + offsetY}px`;
+    });
+
+    element.call(drag);
+}
+
+// Default behaviour for the legend is to remain pinned to the top right hand corner with a specific margin.
+// Once the legend has moved we cannot continue to use that css based approach.
+function overrideDefaultResizingBehaviour(node) {
+    d3.select(window).on("resize", function() {
```

Comment:
From a d3 point of view, it would be better to namespace this event handler, because "resize" would replace any other "resize" event registered on the window. e.g.

```javascript
d3.select(window).on("resize.for-dragging", function() {
```

Reply:
As we also redraw the chart (including the legend) when it resizes (both for window resize and the content area resizing), it may not be necessary to handle the resize event here at all. i.e. the top level function `enableDragging(element)` should be called on an existing legend element, so in principle you should just be able to check that it's still on the screen there.

Reply:
Alrighty, name spacing the event is something new to me, but done. Makes sense.

I want to stick with the event in general, because - having tried it - just enforcing the legend's location being onscreen after the redraw doesn't look great; everything else gradually crunches or stretches, and then the legend just appears after-the-fact.

I've added an boolean attribute to the legend-container: pinned. The legend becomes pinned whenever it is in the top right corner, and returns to using the default resizing behaviour. The event handler is only set and unset when this bool flips. Let me know what you think of the changes.

Reply:
Also, ideally the legend would animate in an obvious way when it pins or unpins, but I'd say that's a later thing.